### PR TITLE
CB2-5377 Look up vehicle details from Atos VEHICLE table.

### PR DESCRIPTION
**Pull request by Ian-Hodges**
## Description

Some vehicle details are not readily available and need to be fetched from the VEHICLE table in Atos.

Related issue: [CB2-5377](https://dvsa.atlassian.net/browse/CB2-5377)

## Evidence of Completion
Details of the vehicle with vrm of AB12CDE.
image

After running the function a new row is added to the VEHICLE_BOOKING table.
image

Amend the vehicle with vrm of AB12CDE.
image

Function ran again. A new row is added which includes the updated value for the vehicle.
image

### CODE REVIEW

File: src/database/vehicleBookingDb.ts [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+export const vehicleBookingDb = {
+  async insert(vehicleBooking: VehicleBooking): Promise<void> {
+    const connection = await dbConnection();
+
+    vehicleBooking.FK_BKGHDR_NO = connection.raw(
+      "COALESCE((SELECT MAX(FK_BKGHDR_NO) + 1 from VEHICLE_BOOKING where FK_BKGHDR_USER_NO = 'XR'), 1)",
+    );
+
+    const insertResult: string[] = await connection
```
Comment (tahayazici-bjss) [inline src/database/vehicleBookingDb.ts]: `insertResults`? If you are expecting an array of strings.

File: src/vehicleBooking/vehicleBooking.ts [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+    booking.SHORT_NAME = vtBooking.name;
+    booking.VEHICLE_CLASS = vehicle.VEHICLE_CLASS;
+    booking.NO_OF_AXLES = vehicle.NO_OF_AXLES;
+    booking.TIMESTAMP0 = new Date(vtBooking.bookingDate);
+    booking.VRM = vtBooking.vrm;
+    booking.FK_APPTYP_APPL_TYP = vtBooking.testCode;
+    booking.FK_LANTBD_DATE = new Date(vtBooking.testDate);
+    booking.FK_STATN_ID = vtBooking.pNumber;
+    booking.FK_VEH_SYST_NO = vehicle.SYSTEM_NUMBER;
+    booking.COUNTED_AXLES = vehicle.NO_OF_AXLES;
```
Comment (tahayazici-bjss) [inline src/vehicleBooking/vehicleBooking.ts]:
```suggestion
    const booking = {
      ...(new VehicleBooking()),
      SHORT_NAME: vtBooking.name,
      VEHICLE_CLASS: vehicle.VEHICLE_CLASS,
      NO_OF_AXLES: vehicle.NO_OF_AXLES,
      TIMESTAMP0: new Date(vtBooking.bookingDate),
      VRM: vtBooking.vrm,
      FK_APPTYP_APPL_TYP: vtBooking.testCode,
      FK_LANTBD_DATE: new Date(vtBooking.testDate),
      FK_STATN_ID: vtBooking.pNumber,
      FK_VEH_SYST_NO: vehicle.SYSTEM_NUMBER,
      COUNTED_AXLES: vehicle.NO_OF_AXLES
    };
```

File: src/database/vehicleDb.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { dbConnection } from './dbConnection';
+import { Vehicle } from '../interfaces/Vehicle';
+
+export const vehicleDb = {
+  async get(vrm: string): Promise<Vehicle> {
+    const connection = await dbConnection();
+
+    const getResult: Vehicle[] = await connection
```
Comment (tahayazici-bjss) [inline src/database/vehicleDb.ts]: You are initializing a variable to verify and return later on. It is not a method or a getter so just `results` would be more adequate.

File: src/database/vehicleBookingDb.ts (hunk 1 above)
Comment (Ian-Hodges) [reply src/database/vehicleBookingDb.ts]: Will amend to results.

File: src/database/vehicleDb.ts (hunk 3 above)
Comment (Ian-Hodges) [reply src/database/vehicleDb.ts]: Will amend to results.

File: src/vehicleBooking/vehicleBooking.ts (hunk 2 above)
Comment (Ian-Hodges) [reply src/vehicleBooking/vehicleBooking.ts]: I'm not a fan of this in C#. If one of the assignments has an issue, it's not clear in the stack trace which one was at fault. I'll see if the same issue occurs in JavaScript and amend if it doesn't.

Comment (Ian-Hodges) [reply src/vehicleBooking/vehicleBooking.ts]: It's not an issue in Javascript.

Comment (tahayazici-bjss) [reply src/vehicleBooking/vehicleBooking.ts]: Thanks for looking into it mate 👍

**Review by tahayazici-bjss — APPROVED** [review]

File: src/database/vehicleDb.ts [hunk 4]
```diff
@@ -0,0 +1,23 @@
+import { dbConnection } from './dbConnection';
+import { Vehicle } from '../interfaces/Vehicle';
+
+export const vehicleDb = {
+  async get(vrm: string): Promise<Vehicle> {
```
Comment (TheWorldWideWebb) [inline src/database/vehicleDb.ts:5]: If this is the function that gets those fields for the main query it won't want to be Async

Comment (Ian-Hodges) [reply src/database/vehicleDb.ts:5]: Where it's called, it's awaited so we have the result before it moves on to the insert.
image

File: src/database/vehicleDb.ts [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+    const connection = await dbConnection();
+
+    const results: Vehicle[] = await connection
+      .select('VEHICLE_CLASS', 'NO_OF_AXLES', 'SYSTEM_NUMBER')
+      .from<Vehicle>('VEHICLE')
+      .where('CURR_REGMK', vrm.padEnd(8, ' '));
```
Comment (aaron-downing) [inline src/database/vehicleDb.ts:11]: Are we also padding the VRMs we're inserting into VEHICLE_BOOKING?

File: src/database/vehicleDb.ts (hunk 4 above)
Comment (ghost) [reply src/database/vehicleDb.ts:5]: Thanks, I was looking for this line but couldn't find it

File: src/database/vehicleDb.ts (hunk 5 above)
Comment (Ian-Hodges) [reply src/database/vehicleDb.ts:11]: There's no need to pad when inserting. The VEHICLE_BOOKING.VRM column is VARCHAR2(8).

The VEHICLE.CURR_REGMK column is a CHAR(8) so we have to pad when searching. Oracle is supposed to be clever and match when the values differ only in the number of trailing blanks. However, locally it wasn't working for me.

**Review by aaron-downing — APPROVED** [review]

**Pull request merged**